src/jit/IntMath32Inl.h

```diff
@@ -1206,3 +1206,164 @@ static void emitConvert(sljit_compiler* compiler, Instruction* instr)
     }
     }
 }
+
+void emitMemory(sljit_compiler* compiler, Instruction* instr)
+{
+    Operand* operands = instr->operands();
+
+    CompileContext* context = CompileContext::get(compiler);
+    Walrus::Memory* memory = context->compiler->memories().at(instr->value().memory.memidx);
+
+    sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(kContextReg),
+                   OffsetOfContextField(memory0Ptr), SLJIT_IMM,
+                   reinterpret_cast<sljit_sw>(memory->buffer()));
+    sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(kContextReg),
+                   OffsetOfContextField(memory0Current), SLJIT_IMM,
+                   memory->sizeInPageSize());
+
+    sljit_sw memType;
+    struct sljit_jump* cmp;
+
+    if (((operands + 1)->location.valueInfo & LocationInfo::kSizeMask) == 1) {
+        JITArg args[2];
+
+        for (Index i = 0; i < (instr->paramCount() + instr->resultCount()); ++i) {
+            operandToArg(operands, args[i]);
+            operands++;
+        }
+
+        switch (instr->opcode()) {
+        case Walrus::I32Load8SOpcode:
+            memType = SLJIT_MOV32_S8;
+            break;
+        case Walrus::I32Store8Opcode:
+        case Walrus::I32Load8UOpcode:
+            memType = SLJIT_MOV32_U8;
+            break;
+        case Walrus::I32Load16SOpcode:
+            memType = SLJIT_MOV32_S16;
+            break;
+        case Walrus::I32Load16UOpcode:
+        case Walrus::I32Store16Opcode:
+            memType = SLJIT_MOV32_U16;
+            break;
+        case Walrus::I32LoadOpcode:
+        case Walrus::I32StoreOpcode:
+            memType = SLJIT_MOV32;
+            break;
+        default:
+            WABT_UNREACHABLE;
+            return;
+        }
+        if (args[0].arg != SLJIT_IMM) {
+            sljit_emit_op2(compiler, SLJIT_ADD32, SLJIT_R0, 0, SLJIT_MEM1(args[0].arg), args[0].argw, SLJIT_IMM, 4);
+            sljit_emit_op2(compiler, SLJIT_ADD32, SLJIT_R1, 0, SLJIT_MEM1(args[0].arg), args[0].argw, SLJIT_IMM, instr->value().memory.offset);
+        } else {
+            sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R0, 0, args[0].arg, args[0].argw, SLJIT_IMM, 4);
+            sljit_emit_op2(compiler, SLJIT_ADD32, SLJIT_R1, 0, args[0].arg, args[0].argw, SLJIT_IMM, instr->value().memory.offset);
+        }
+
+        sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R0, 0, SLJIT_R0, 0, SLJIT_IMM, instr->value().memory.offset);
+        cmp = sljit_emit_cmp(compiler, SLJIT_LESS, SLJIT_R0, 0, SLJIT_IMM, memory->sizeInByte());
```

Comment:
The problem with this code is memory offset / size can be changed in runtime, leads to a crash.

There is one more interesting problem here. If offset is high, then adding memory start might cause an unsigned overflow, and the comparison here might pass even if it fails.